### PR TITLE
[enhancement] Implement Swappable models

### LINKED ISSUE #45: [enhancement] Implement Swappable models
For implementing [swappable models](https://github.com/wq/django-swappable-models) check [django-freeradius](https://github.com/openwisp/django-freeradius/blob/master/django_freeradius/models.py) as a reference.

This issue can be broken down to:-
* Create ​Abstract Models​ from existing models in a base directory.
* Import the newly created abstract models in respective apps and apply `​swappable` models​ followed by  updating migration files and adding the dependency of swapper.
* Add new tests and document​ this to highlight to the end-user about the reusability of this app.

### PR DESCRIPTION
**Pull request by nepython**
closes #45
This is still WIP and I have implemented it only across `monitoring` app. There were few queries like `proxy` model and `abstract` model not compatible with each other which would be needed to create `AbstactDeviceData`.
**NOTE**: In the `AbstractMetric` model, I was unable to define `full_clean` method as it would lead to recursive calls when models are being loaded and thus defined it in the `Metric` model.

### CODE REVIEW

**Comment by nepython** [comment]
Implemented swappable models across `check`, `monitoring` and `device` apps. I have added `AbstractAdmin` too. Though it might not be required in order to implement swappable models, I found it being done in other openwisp modules while implementing abstract models (like [here](https://github.com/openwisp/django-freeradius/pull/13/files#diff-1216370c29a0f43ad4b424c35aab2f1a)) so followed the same.

**Warning**: As mentioned above, I have not implemented it for `DeviceData` which being a `proxy` model is imcompatible with `abstract` model.

File: README.rst [hunk 1]
```diff
@@ -467,3 +467,94 @@ Run tests with:
 .. code-block:: shell
 
     ./runtests.py
+
+The django apps in ``tests/openwisp2/`` namely ``sample_monitoring``, ``sample_check``, ``sample_device_monitoring`` adds some changes on
```
Comment (PabloCastellano) [inline README.rst:471]: typo: adds -> add

File: openwisp_monitoring/monitoring/base/admin.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+        (None, {'fields': ('name', 'description', 'content_type', 'object_id',)}),
+        (
+            _('Advanced options'),
+            {'classes': ('collapse',), 'fields': ('key', 'field_name')},
+        ),
+    )
+
+    class Media:
+        css = {'all': ('monitoring/css/monitoring.css',)}
+        js = ('monitoring/js/plotly.min.js', 'monitoring/js/graph.js')
```
Comment (PabloCastellano) [inline openwisp_monitoring/monitoring/base/admin.py:37]: Why did you change from `monitoring/js/plotly-cartesian.min.js` to `monitoring/js/plotly.min.js`?

File: openwisp_monitoring/device/base/tests/test_recovery.py [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+from django.urls import reverse
+from swapper import load_model
+
+from ....check.tasks import auto_create_ping
+from ...models import DeviceMonitoring
+from ...signals import health_status_changed
+from ...tasks import trigger_device_checks
+from ...utils import get_device_recovery_cache_key
+
+# TODO: this is not yet extensible
```
Comment (PabloCastellano) [inline openwisp_monitoring/device/base/tests/test_recovery.py:14]: What do you mean?

File: openwisp_monitoring/monitoring/base/admin.py (hunk 2 above)
Comment (nepython) [reply openwisp_monitoring/monitoring/base/admin.py:37]: Didn't notice it, I had made changes to it prior to `Charts` getting refactored and thus this change (not sure) might have been introduced then.

File: README.rst [hunk 4]
```diff
@@ -466,4 +466,108 @@ Run tests with:
...
+
+All apps in *openwisp-monitoring* provide a set of models and admin classes which can
+be imported, extended and reused by third party apps.
+
+To extend any app in *openwisp-monitoring*, you **MUST NOT** add it to ``settings.INSTALLED_APPS``,
+but you must create your own app (which goes into ``settings.INSTALLED_APPS``), import the
+base classes from the respective app in ``openwisp-monitoring`` and add your customizations.
+
+In order to help django find the static files and templates of *openwisp-monitoring*,
+you need to perform the steps described below.
```
Comment (nemesisdesign) [inline README.rst:502]: Don't forget the premise mentioned here: https://github.com/openwisp/openwisp-firmware-upgrader#extending-openwisp-firmware-upgrader

File: README.rst [hunk 5]
```diff
@@ -466,4 +466,108 @@ Run tests with:
...
+    EXTENDED_APPS = ('monitoring', 'check', 'device_monitoring')
+
+3. Add ``openwisp_utils.staticfiles.DependencyFinder``
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Add ``openwisp_utils.staticfiles.DependencyFinder`` to
+``STATICFILES_FINDERS`` in your ``settings.py``:
+
+.. code-block:: python
+    STATICFILES_FINDERS = [
```
Comment (nemesisdesign) [inline README.rst:526]: please leave a blank line after `.. code-block:: python`

File: README.rst [hunk 6]
```diff
@@ -466,4 +466,108 @@ Run tests with:
...
+        'openwisp_utils.staticfiles.DependencyFinder',
+    ]
+
+4. Add ``openwisp_utils.loaders.DependencyLoader``
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Add ``openwisp_utils.loaders.DependencyLoader`` to ``TEMPLATES`` in your ``settings.py``:
+
+.. code-block:: python
+    TEMPLATES = [
```
Comment (nemesisdesign) [inline README.rst:538]: same here

File: README.rst [hunk 7]
```diff
@@ -466,4 +466,108 @@ Run tests with:
...
+
+    pip install --U https://github.com/openwisp/openwisp-controller/tarball/master
+
+2. Add ``EXTENDED_APPS``
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+Add the following to your ``settings.py``:
+
+.. code-block:: python
+    EXTENDED_APPS = ('monitoring', 'check', 'device_monitoring')
```
Comment (nemesisdesign) [inline README.rst:517]: same here

File: openwisp_monitoring/check/base/tests/test_models.py [hunk 8]
```diff
@@ -0,0 +1,55 @@
+from django.core.exceptions import ValidationError
+from openwisp_monitoring.check.classes import Ping
+from openwisp_monitoring.check.settings import CHECK_CLASSES
+
+
```
Comment (nemesisdesign) [inline openwisp_monitoring/check/base/tests/test_models.py:5]: I just had an idea.. maybe dividing the test classes like we're doing here is not necessary.
Can you test what happens if we keep the tests as they are with the only difference that we load the models using swappable.load_model instead of importing directly?

It should work! If it does we can avoid quite some hassle.

File: tests/openwisp2/sample_device_monitoring/admin.py [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+
+DeviceAdmin.list_display.insert(
+    DeviceAdmin.list_display.index('config_status'), 'health_status'
+)
+DeviceAdmin.list_select_related += ('monitoring',)
+DeviceAdmin.list_filter.insert(
+    0, 'monitoring__status',
+)
+DeviceAdmin.fields.insert(DeviceAdmin.fields.index('last_ip'), 'health_status')
+DeviceAdmin.readonly_fields.append('health_status')
```
Comment (nemesisdesign) [inline tests/openwisp2/sample_device_monitoring/admin.py:43]: as a general rule, avoid requiring the duplication of so many lines of code. The way I showed in https://github.com/openwisp/openwisp-firmware-upgrader/pull/60/ will help you to solve this.

File: tests/openwisp2/sample_device_monitoring/models.py [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+            [monitoring.is_metric_critical(metric), related_status == 'critical']
+        ):
+            status = 'critical'
+        monitoring.update_status(status)
+
+
+@receiver(post_save, sender=Device, dispatch_uid='create_device_monitoring')
+def device_monitoring_receiver(sender, instance, created, **kwargs):
+    if created:
+        DeviceMonitoring.objects.create(device=instance)
```
Comment (nemesisdesign) [inline tests/openwisp2/sample_device_monitoring/models.py:49]: we have to find a way to avoid all this duplication, this should be in the base classes.
Try loading `DeviceMonitoring` using `swappable.load_model`

File: tests/openwisp2/sample_monitoring/admin.py [hunk 11]
```diff
@@ -0,0 +1,21 @@
+from django.contrib import admin
+from openwisp_monitoring.monitoring.base.admin import (
+    AbstracThresholdAdmin,
+    AbstractMetricAdmin,
+)
+from swapper import load_model
+
+Metric = load_model("sample_monitoring", "Metric")
+Threshold = load_model("sample_monitoring", "Threshold")
```
Comment (nemesisdesign) [inline tests/openwisp2/sample_monitoring/admin.py:9]: use single quotes for consistency, eg: `'sample_monitoring'`

File: tests/openwisp2/sample_monitoring/models.py [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+    AbstractMetric,
+    AbstractThreshold,
+)
+
+
+class Metric(AbstractMetric):
+    def full_clean(self, *args, **kwargs):
+        # clean up key before field validation
+        self.key = self._makekey(self.key)
+        return super(Metric, self).full_clean(*args, **kwargs)
```
Comment (nemesisdesign) [inline tests/openwisp2/sample_monitoring/models.py:12]: Did you add this? What makes it necessary?

File: tests/openwisp2/sample_check/models.py [hunk 13]
```diff
@@ -0,0 +1,8 @@
+from openwisp_monitoring.check.base.models import AbstractCheck
+from swapper import swappable_setting
+
+
+class Check(AbstractCheck):
```
Comment (nemesisdesign) [inline tests/openwisp2/sample_check/models.py:5]: Take a look at this: https://github.com/openwisp/openwisp-firmware-upgrader/blob/master/tests/openwisp2/sample_firmware_upgrader/models.py#L12-L16

It's a good way to test the extensibility.
We just have to ensure the field is shown in the admin with an extra test for each admin class.

Apply this to all models please.

File: README.rst [hunk 14]
```diff
@@ -466,4 +466,115 @@ Run tests with:
...
+                ],
+                'context_processors': [
+                    'django.template.context_processors.debug',
+                    'django.template.context_processors.request',
+                    'django.contrib.auth.context_processors.auth',
+                    'django.contrib.messages.context_processors.messages',
+                ],
+            },
+        }
+    ]
```
Comment (nemesisdesign) [inline README.rst:562]: I think this is missing:

[5. Add swapper configurations](https://github.com/openwisp/openwisp-firmware-upgrader#5-add-swapper-configurations)

File: README.rst [hunk 15]
```diff
@@ -466,4 +466,115 @@ Run tests with:
...
+To extend ``device_monitoring`` app, please checkout the `sample_device_monitoring models.py file <https://github.com/openwisp/openwisp-monitoring/tree/master/tests/openwisp2/sample_device_monitoring/models.py>`_.
+
+Extending the admin
+~~~~~~~~~~~~~~~~~~~
+
+To extend ``check`` app, please checkout the `sample_check admin.py file <https://github.com/openwisp/openwisp-monitoring/tree/master/tests/openwisp2/sample_check/admin.py>`_.
+
+To extend ``monitoring`` app, please checkout the `sample_monitoring admin.py file <https://github.com/openwisp/openwisp-monitoring/tree/master/tests/openwisp2/sample_monitoring/admin.py>`_.
+
+To extend ``device_monitoring`` app, please checkout the `sample_device_monitoring admin.py file <https://github.com/openwisp/openwisp-monitoring/tree/master/tests/openwisp2/sample_device_monitoring/admin.py>`_.
```
Comment (nemesisdesign) [inline README.rst:580]: I think this is missing: [Reusing the base tests](https://github.com/openwisp/openwisp-firmware-upgrader#reusing-the-base-tests)

File: tests/openwisp2/sample_monitoring/models.py (hunk 12 above)
Comment (nemesisdesign) [reply tests/openwisp2/sample_monitoring/models.py:12]: I just read now that this is giving you issues. Next time you have some issues in the code, leave us an inline comment, it's easier to review and discuss.

The line which is giving you issues may be:

```python
return super(Metric, self).full_clean(*args, **kwargs)
```

Change this to:

```python
return super().full_clean(*args, **kwargs)
```

Please take advantage of this chance to look for all occurrences of `super(` in this project and make it consistent, that is, switch to this way of calling super, which is available only in python 3 (while when I wrote this module it had to support python 2.7 too).

File: openwisp_monitoring/check/tasks.py [hunk 16]
```diff
@@ -38,6 +39,7 @@ def auto_create_ping(model, app_label, object_id, created):
     """
     Called by openwisp_monitoring.check.models.auto_ping_receiver
     """
+    Check = load_model('check', 'Check')
```
[the remark this answers is not in the document]
Comment (nepython) [reply openwisp_monitoring/check/tasks.py:42]:
```python
File "/home/hardik/openwisp/openwisp-monitoring/openwisp_monitoring/check/base/models.py", line 81, in <module>
    from openwisp_monitoring.check.tasks import auto_create_ping
```

Can't do it.

File: tests/openwisp2/sample_monitoring/models.py [hunk 17]
```diff
@@ -0,0 +1,22 @@
...
+    AbstractThreshold,
+)
+
+
+class DetailsModel(models.Model):
+    details = models.CharField(max_length=64, blank=True, null=True)
+
+
+class Metric(AbstractMetric):
+    pass
```
Comment (nemesisdesign) [inline tests/openwisp2/sample_monitoring/models.py:14]: see the changes in this commit and replicate them please: https://github.com/openwisp/openwisp-firmware-upgrader/pull/60/commits/604e04491ce9dc0b1108b6b39c7079f540d7ea57#diff-28e2ae300b95f10c04fcf32b14819f36R17-R46

File: openwisp_monitoring/notifications/tests/test_admin.py [hunk 18]
```diff
@@ -27,6 +30,7 @@ def pk(self):
 request.user = MockSuperUser()
 
 
+@skipIf(is_swapped('monitoring', 'Threshold'), 'Running tests on sample_app')
```
[the remark this answers is not in the document]
Comment (nepython) [reply openwisp_monitoring/notifications/tests/test_admin.py:33]: `notifications` app is removed anyways :sweat_smile:

File: openwisp_monitoring/device/tests/test_recovery.py [hunk 19]
```diff
@@ -26,6 +29,7 @@ def _create_device_monitoring(self):
         dm.save()
         return dm
 
+    @skipIf(is_swapped('monitoring', 'Metric'), 'Running tests on sample_app')
```
[the remark this answers is not in the document]
Comment (nepython) [reply openwisp_monitoring/device/tests/test_recovery.py:32]: I am using `swapper.load_model` almost everywhere but still few tests are failing. I will try tweaking them if there aren't many changes, skipped them for the moment.

**NOTE**: Tests in ``SAMPLE_APP`` are running twice over, I am not sure why but this wasn't happening previously when I was running `tests` directly

**Comment by nepython** [comment]
Reposting, what I wrote earlier on chat,

**NOTE**: I have used proxy models for device_monitoring app as it was the best thing I could do to avoid duplication for `DeviceMonitoring` and allow extension for `DeviceData` too.

If anyone has any better suggestions, I am up for it. I tried many workarounds but they didn't work, the only downside of this might be that device_monitoring app needs to be in **INSTALLED_APPS** besides that I didn't find any until now atleast.

**Comment by nepython** [comment]
@nemesisdesign, the tests are running twice in sample_app tests, I checked it out to find that tests are running once the test_classes are imported and second time when the new test_classes derive from original ones. I am unsure why so, since it didn't occur in `firmware_upgrader`.

Secondly, I tried upon your suggestion to remove `skipIf` from all places but couldn't do so for two instances. The `test_api` and for few tests in `check` app making use of `run_checks` command. I think it can be removed but will only involve making new `urls.py` file and tweaking many tests. Though I am not sure how to do it for `run_checks`. Let me know if that needs to done because in my opinion it would mean too many modifications for sample_app, hence I haven't done so.

**Comment by nemesisdesign** [comment]
This is not good enough. I sent you some feedback via IM and left some details instructions here: https://github.com/openwisp/openwisp-monitoring/pull/102.
Please go ahead on that branch and follow the instructions in the PR closely.
